monsterCatalog.py

```diff
@@ -188,180 +190,143 @@ def monsters():
+    }
+    fb_token_exchange_result = requests.get(fb_token_exchange_url,
+                                            params=fb_token_exchange_params)
+    fb_token_exchange_text = fb_token_exchange_result.text
+    long_lived_access_token = fb_token_exchange_text.split("&")[0].split('=')[1]
+    return long_lived_access_token
+
+
+def get_fb_user_data(long_lived_access_token):
+    """Using long-lived access token, query Facebook Graph API for user data."""
+    fb_user_query_url = 'https://graph.facebook.com/v2.4/me'
+    fb_user_query_params = {
+        'fields': 'name,id,email',
+        'access_token': long_lived_access_token
+    }
+    fb_user_query_result = requests.get(fb_user_query_url,
+                                        params=fb_user_query_params)
+    fb_user_data = json.loads(fb_user_query_result.text)
+
+    return fb_user_data
+
+
+def add_or_verify_user(user_data):
+    user_id = get_user_id_by_email(user_data['email'])
     if not user_id:
-        user_id = createUser(flask.session)
+        user_id = create_user(user_data)
     flask.session['user_id'] = user_id
-
-    output = ''
-    output += '<h1>Welcome, '
-    output += flask.session['username']
-
-    output += '!</h1>'
-    output += '<img src="'
-    output += flask.session['picture']
-    output += ' " style = "width: 300px; height: 300px;border-radius: 150px;-webkit-border-radius: 150px;-moz-border-radius: 150px;"> '
-
-    flask.flash("Now logged in as %s" % flask.session['username'])
-    return output
+    return user_id
 
 
 @app.route('/fbdisconnect')
 def fbdisconnect():
     facebook_id = flask.session['facebook_id']
-    # The access token must me included to successfully logout
     access_token = flask.session['access_token']
-    url = 'https://graph.facebook.com/%s/permissions?access_token=%s' % (facebook_id, access_token)
-    h = httplib2.Http()
-    result = h.request(url, 'DELETE')[1]
-    return "you have been logged out"
-
-
-def createUser(session):
-    newUser = User(name=session['username'], email=session[
-                   'email'], picture=session['picture'])
-    db_session.add(newUser)
-    db_session.commit()
-    user = db_session.query(User).filter_by(email=session['email']).one()
+
+    logout_params = {
+        'access_token': access_token
+    }
+
+    fb_logout_url = 'https://graph.facebook.com/%s/permissions' % facebook_id
+    logout_result = requests.delete(fb_logout_url,
+                                    params=logout_params)
+    flask.session.clear()
+    return flask.redirect('/')
+
+
+def create_user(user_data):
+    user = User(
+        name=user_data['username'],
+        email=user_data['email'],
+        picture=user_data['data']['url']
+        )
+    try:
+        db_session.add(user)
+        db_session.commit()
+    except sqlalchemy.ext.SQLAlchemyError:
+        db_session.rollback()
+    user = db_session.query(User).filter_by(email=user_data['email']).one()
```

Comment:
How do you want to handle it if this query fails?  I know you just added the user, but maybe you had to rollback, or something.